### PR TITLE
Add support for CosmosSerializer and CosmosSerializationOptions in the Cosmos Provider

### LINKED ISSUE #20670: Add Support for CosmosSerializationOptions in the Cosmos Provider
Enabling features like indented JSON documents and ignoring a property when null. By setting "IgnoreNullValues" to true, for example, properties will not be serialized when null (as opposed to serializing the property with a value of null).

When configuring the DbContext, allow configuration options based on CosmosSerializationOptions. Currently these are:

IgnoreNullValues
Indented
PropertyNamingPolicy

https://docs.microsoft.com/dotnet/api/microsoft.azure.cosmos.cosmosserializationoptions

### PR DESCRIPTION
**Pull request by Marusyk**
Added support for [CosmosSerializationOptions](https://docs.microsoft.com/uk-ua/dotnet/api/microsoft.azure.cosmos.cosmosserializationoptions?view=azure-dotnet) and [CosmosSerializer ](https://docs.microsoft.com/uk-ua/dotnet/api/microsoft.azure.cosmos.cosmosserializer?view=azure-dotnet)

It is a very useful feature when we want to ignore null values when saving data in CosmosDb with Entity Framework
https://stackoverflow.com/questions/66424039/ignore-null-values-when-saving-data-in-cosmosdb-with-entity-framework

Fixes #20670

Please review
Thank you in advance

### CODE REVIEW

File: src/EFCore.Cosmos/Infrastructure/Internal/CosmosDbOptionExtension.cs [hunk 1]
```diff
@@ -560,6 +621,8 @@ public override long GetServiceProviderHashCode()
                     hashCode = (hashCode * 131) ^ (Extension._gatewayModeMaxConnectionLimit?.GetHashCode() ?? 0);
                     hashCode = (hashCode * 397) ^ (Extension._maxTcpConnectionsPerEndpoint?.GetHashCode() ?? 0);
                     hashCode = (hashCode * 131) ^ (Extension._maxRequestsPerTcpConnection?.GetHashCode() ?? 0);
+                    hashCode = (hashCode * 131) ^ (Extension._serializer?.GetHashCode() ?? 0);
```
Comment (Marusyk) [inline src/EFCore.Cosmos/Infrastructure/Internal/CosmosDbOptionExtension.cs:624]: not sure about this code. can anyone explain how to implement it in a proper way?

File: test/EFCore.Cosmos.FunctionalTests/ConfigPatternsCosmosTest.cs [hunk 2]
```diff
@@ -127,6 +127,34 @@ public async Task Should_throw_if_specified_connection_mode_is_wrong()
                 });
         }
 
+        [ConditionalFact]
+        public async Task Should_use_serialization_options()
```
Comment (Marusyk) [inline test/EFCore.Cosmos.FunctionalTests/ConfigPatternsCosmosTest.cs:131]: It is just for example. I actually don't know where I should place tests for this.
Also, after `SaveChanges`, the `Name` will be stored with `null`. It seems I missed something. Could you please help?

**Review by Marusyk — COMMENTED** [review]
Need to add more tests and fix the behavior.

**Comment by AndriySvyryd** [comment]
`CosmosSerializationOptions` and `CosmosSerializer` shouldn't be exposed on any public API as they will move to `Azure.Cosmos.Serialization` in SDK v4, see: [CosmosSerializer.cs](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/d706a6d4421e6e4ecbb112732f8f24363d9640e1/Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializer.cs#L19), [CosmosSerializationOptions.cs](https://github.com/Azure/azure-cosmos-dotnet-v3/blob/d706a6d4421e6e4ecbb112732f8f24363d9640e1/Microsoft.Azure.Cosmos/src/Serializer/CosmosSerializationOptions.cs)

#20670 and #17306 should be implemented after #18753 and #14570 are done

**Comment by Marusyk** [comment]
Hi @AndriySvyryd
Is it means that we should wait for SDK v4 release (November 2021) after that, we can implement this feature in EF and release it in November 2022? 😮

**Comment by AndriySvyryd** [comment]
@Marusyk We can upgrade to a feature complete preview of V4, this should give us enough time before release

**Comment by AndriySvyryd** [comment]
In any case a better approach would be to expose the options on OptionsBuilder directly

**Pull request closed without merging**